File: src/mechanics/combat.rs
/// Trait shared by all combatants (players, enemies, etc.).
pub trait Combatant {
    fn name(&self) -> &str;
    fn hp(&self) -> i32;
    fn max_hp(&self) -> i32;
    fn take_damage(&mut self, amount: i32);

    fn is_alive(&self) -> bool {
        self.hp() > 0
    }

    fn display_status(&self) -> String {
        format!("{}: {}/{} HP", self.name(), self.hp(), self.max_hp())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    struct DummyCombatant {
        name: String,
        hp: i32,
        max_hp: i32,
    }

    impl Combatant for DummyCombatant {
        fn name(&self) -> &str {
            &self.name
        }
        fn hp(&self) -> i32 {
            self.hp
        }
        fn max_hp(&self) -> i32 {
            self.max_hp
        }
        fn take_damage(&mut self, amount: i32) {
            self.hp = (self.hp - amount).max(0);
        }
    }

    #[test]
    fn take_damage_reduces_hp() {
        let mut e = DummyCombatant {
            name: "T".into(),
            hp: 3,
            max_hp: 3,
        };
        e.take_damage(1);
        assert_eq!(e.hp(), 2);
        assert!(e.is_alive());
    }

    #[test]
    fn damage_does_not_go_below_zero() {
        let mut e = DummyCombatant {
            name: "T".into(),
            hp: 1,
            max_hp: 3,
        };
        e.take_damage(5);
        assert_eq!(e.hp(), 0);
        assert!(!e.is_alive());
    }

    #[test]
    fn display_status_format() {
        let e = DummyCombatant {
            name: "勇者".into(),
            hp: 2,
            max_hp: 3,
        };
        assert_eq!(e.display_status(), "勇者: 2/3 HP");
    }
}

File: docs/js/mechanics/combat.js
/** Base class for all combatants (players, enemies, etc.). */
export class Combatant {
    constructor(name, maxHp, speed = 3) {
        this.name = name;
        this.maxHp = maxHp;
        this.hp = maxHp;
        this.speed = speed;
        this.shield = 0;
    }

    /** Deals damage, absorbing through shield first. */
    takeDamage(amount) {
        const absorbed = Math.min(amount, this.shield);
        this.shield -= absorbed;
        this.hp = Math.max(0, this.hp - (amount - absorbed));
    }

    addShield(amount) {
        this.shield += amount;
    }

    clearShield() {
        this.shield = 0;
    }

    /** Heal HP, capped at maxHp. Returns actual amount healed. */
    heal(amount) {
        const before = this.hp;
        this.hp = Math.min(this.maxHp, this.hp + amount);
        return this.hp - before;
    }

    isAlive() {
        return this.hp > 0;
    }

    hpPercent() {
        return (this.hp / this.maxHp) * 100;
    }
}
